From 64237903224aeadd04d20b5e34a96da1b7a23442 Mon Sep 17 00:00:00 2001
From: Yongqiang Sun <yongqiang.sun@amd.com>
Date: Mon, 16 May 2016 10:04:12 -0400
Subject: [PATCH 0489/1722] drm/amd/dal: Check edid for HDMI monitor.

Issue:
HDMI monitor without audio not light up.
Cause:
Monitor is considered as a HDMI-DVI dongle,
Signal is donwngraded to singgle link DVI with incorrect color depth
and failed in validate dvi output.
Solution:
Check HDMI monitor by parse edid CEA block.

Signed-off-by: Yongqiang Sun <yongqiang.sun@amd.com>
Acked-by: Jordan Lazare <Jordan.Lazare@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 .../gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_helpers.c  |  3 +++
 drivers/gpu/drm/amd/dal/dc/core/dc_link.c          |  5 +++++
 drivers/gpu/drm/amd/dal/dc/core/dc_resource.c      | 25 ++++++----------------
 drivers/gpu/drm/amd/dal/dc/dc_types.h              |  4 +++-
 4 files changed, 17 insertions(+), 20 deletions(-)

diff --git a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_helpers.c b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_helpers.c
index 4173c1f..6190a51 100644
--- a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_helpers.c
+++ b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_helpers.c
@@ -101,6 +101,9 @@ enum dc_edid_status dm_helpers_parse_edid_caps(
 		}
 	}
 
+	edid_caps->edid_hdmi = drm_detect_hdmi_monitor(
+			(struct edid *) edid->raw_edid);
+
 	sad_count = drm_edid_to_sad((struct edid *) edid->raw_edid, &sads);
 	if (sad_count <= 0) {
 		DRM_INFO("SADs count is: %d, don't need to read it\n",
diff --git a/drivers/gpu/drm/amd/dal/dc/core/dc_link.c b/drivers/gpu/drm/amd/dal/dc/core/dc_link.c
index 5833eaa..8f3fedb 100644
--- a/drivers/gpu/drm/amd/dal/dc/core/dc_link.c
+++ b/drivers/gpu/drm/amd/dal/dc/core/dc_link.c
@@ -687,6 +687,11 @@ bool dc_link_detect(const struct dc_link *dc_link, bool boot)
 			break;
 		}
 
+		/* HDMI-DVI Dongle */
+		if (dc_sink->sink_signal == SIGNAL_TYPE_HDMI_TYPE_A &&
+				!dc_sink->edid_caps.edid_hdmi)
+			dc_sink->sink_signal = SIGNAL_TYPE_DVI_SINGLE_LINK;
+
 		/* Connectivity log: detection */
 		for (i = 0; i < sink->public.dc_edid.length / EDID_BLOCK_SIZE; i++) {
 			CONN_DATA_DETECT(link,
diff --git a/drivers/gpu/drm/amd/dal/dc/core/dc_resource.c b/drivers/gpu/drm/amd/dal/dc/core/dc_resource.c
index 76dcc61..eb3dd59 100644
--- a/drivers/gpu/drm/amd/dal/dc/core/dc_resource.c
+++ b/drivers/gpu/drm/amd/dal/dc/core/dc_resource.c
@@ -689,9 +689,9 @@ static struct audio *find_first_free_audio(struct resource_context *res_ctx)
 
 static void set_stream_signal(struct pipe_ctx *pipe_ctx)
 {
-	struct dc_sink *dc_sink =
-		(struct dc_sink *) pipe_ctx->stream->public.sink;
+	const struct dc_sink *dc_sink = pipe_ctx->stream->public.sink;
 
+	pipe_ctx->signal = dc_sink->sink_signal;
 	/* For asic supports dual link DVI, we should adjust signal type
 	 * based on timing pixel clock. If pixel clock more than 165Mhz,
 	 * signal is dual link, otherwise, single link.
@@ -700,23 +700,10 @@ static void set_stream_signal(struct pipe_ctx *pipe_ctx)
 			dc_sink->sink_signal == SIGNAL_TYPE_DVI_DUAL_LINK) {
 		if (pipe_ctx->stream->public.timing.pix_clk_khz >
 						TMDS_MAX_PIXEL_CLOCK_IN_KHZ)
-			dc_sink->sink_signal = SIGNAL_TYPE_DVI_DUAL_LINK;
+			pipe_ctx->signal = SIGNAL_TYPE_DVI_DUAL_LINK;
 		else
-			dc_sink->sink_signal = SIGNAL_TYPE_DVI_SINGLE_LINK;
+			pipe_ctx->signal = SIGNAL_TYPE_DVI_SINGLE_LINK;
 	}
-
-	pipe_ctx->signal = dc_sink->sink_signal;
-
-	/* Down-grade pipe_ctx signal instead of sink singal from HDMI to DVI
-	 * here based on audio info in stream. This allows DC to handle stream
-	 * with or without audio on a HDMI connector.
-	 *
-	 * On a HDMI to DVI passive dongle, audio info is not available from the
-	 * EDID and the signal is down-graded in pipe ctx.
-	 */
-	if (pipe_ctx->signal == SIGNAL_TYPE_HDMI_TYPE_A &&
-			pipe_ctx->stream->public.audio_info.mode_count == 0)
-		pipe_ctx->signal = SIGNAL_TYPE_DVI_SINGLE_LINK;
 }
 
 bool resource_is_stream_unchanged(
@@ -923,8 +910,8 @@ enum dc_status resource_map_pool_resources(
 
 			/* TODO: Add check if ASIC support and EDID audio */
 			if (!stream->sink->converter_disable_audio &&
-						dc_is_audio_capable_signal(
-						pipe_ctx->signal)) {
+						dc_is_audio_capable_signal(pipe_ctx->signal) &&
+						stream->public.audio_info.mode_count) {
 				pipe_ctx->audio = find_first_free_audio(
 						&context->res_ctx);
 
diff --git a/drivers/gpu/drm/amd/dal/dc/dc_types.h b/drivers/gpu/drm/amd/dal/dc/dc_types.h
index 2702b43..fd76440 100644
--- a/drivers/gpu/drm/amd/dal/dc/dc_types.h
+++ b/drivers/gpu/drm/amd/dal/dc/dc_types.h
@@ -259,7 +259,9 @@ struct dc_edid_caps {
 	uint32_t video_latency;
 
 	/*HDMI 2.0 caps*/
-	uint8_t lte_340mcsc_scramble;
+	bool lte_340mcsc_scramble;
+
+	bool edid_hdmi;
 };
 
 struct view {
-- 
2.7.4

